pedsnetdcc/id_mapping_transform.py

```diff
@@ -0,0 +1,239 @@
+
+        # Execute the new ID mapping count statement and ensure it didn't
+        # error and did return a result.
+        new_id_count_stmt.execute(conn_str)
+        check_stmt_err(new_id_count_stmt, 'ID mapping pre-transform')
+        check_stmt_data(new_id_count_stmt, 'ID mapping pre-transform')
+
+        # Get the actual count of new ID maps needed and log it.
+        results['new_id_count'] = new_id_count_stmt.data[0][0]
+        logger.info({'msg': 'counted new IDs needed', 'table': table_name,
+                     'count': results['new_id_count']})
+
+        # Build list of two last id table update statements that need to occur
+        # in a single transaction to prevent race conditions.
+        update_last_id_stmts = StatementList()
+        update_last_id_stmts.append(Statement(
+            lock_last_id_sql.format(**results),
+            lock_last_id_msg.format(**results)))
+        update_last_id_stmts.append(Statement(
+            update_last_id_sql.format(**results),
+            update_last_id_msg.format(**results)))
+
+        # Execute last id table update statements and ensure they didn't error
+        # and the second one returned results.
+        update_last_id_stmts.serial_execute(conn_str, transaction=True)
+
+        for stmt in update_last_id_stmts:
+            check_stmt_err(stmt, 'ID mapping pre-transform')
+        check_stmt_data(update_last_id_stmts[1], 'ID mapping pre-transform')
+
+        # Get the old and new last IDs from the second update statement.
+        results['old_last_id'] = update_last_id_stmts[1].data[0][0]
+        results['new_last_id'] = update_last_id_stmts[1].data[0][1]
+        logger.info({'msg': 'last ID tracking table updated',
+                     'table': table_name,
+                     'old_last})d': results['old_last_id'],
+                     'new_last_id': results['new_last_id']})
+
+        # Build statement to generate new ID maps.
+        insert_new_maps_stmt = Statement(
+            insert_new_maps_sql.format(**results),
+            insert_new_maps_msg.format(**results))
+
+        # Execute new map generation statement and ensure it didn't error.
+        insert_new_maps_stmt.execute(conn_str)
+        check_stmt_err(insert_new_maps_stmt, 'id mapping pre-transform')
+        logging.info({'msg': 'generated new ID mappings', 'table': table_name,
+                      'count': insert_new_maps_stmt.rowcount,
+                      'elapsed': secs_since(starttime)})
+
+    @classmethod
+    def modify_select(cls, metadata, table_name, select=None, join=None):
+        """Alter foreign key columns to get mapped DCC IDs.
+
+        The primary key and each foreign key which points at a data table
+        (i.e., not a vocabulary table) is altered to instead get the dcc_id's
+        from the mapping table, which requires a join to be added.
+        """
+
+        # Raise error if attempted on a vocabulary table.
+        if table_name in VOCAB_TABLES:
```

Comment:
Ah, if we are leaving open the possibility of transforming the vocabulary tables, then the other existing transformations should have the same feature. I guess I was assuming that the transformation runner would skip vocab tables.